src/middleware/passport-auth.ts

```diff
@@ -0,0 +1,168 @@
+import { createPrivateKey } from 'node:crypto';
+
+import passport from 'passport';
+import { Issuer, Strategy as OpenIdStrategy, TokenSet, UserinfoResponse } from 'openid-client';
+import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
+import { Strategy as JWTStrategy, ExtractJwt } from 'passport-jwt';
+import { Repository } from 'typeorm';
+
+import { logger } from '../utils/logger';
+import { User } from '../entities/user';
+
+const readPrivateKey = (privateKey: string) => {
+    return createPrivateKey({ key: Buffer.from(privateKey, 'base64'), type: 'pkcs8', format: 'der' });
+};
+
+export const initPassport = async (userRepository: Repository<User>): Promise<void> => {
+    passport.use(
+        'jwt',
+        new JWTStrategy(
+            {
+                jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+                secretOrKey: process.env.JWT_SECRET!
+            },
+            async (jwtPayload, done) => {
+                logger.info('request authenticated with JWT');
+                const user = await userRepository.findOneBy({ id: jwtPayload?.user?.id });
+                return done(null, user);
+            }
+        )
+    );
+
+    if (process.env.AUTH_PROVIDERS?.includes('onelogin')) {
+        const oneLoginIssuer = await Issuer.discover(`${process.env.ONELOGIN_URL}/.well-known/openid-configuration`);
+        const privateKey = process.env.ONELOGIN_PRIVATE_KEY!.replace(/\\n/g, '\n');
+
+        passport.use(
+            'onelogin',
+            new OpenIdStrategy(
+                {
+                    client: new oneLoginIssuer.Client(
+                        {
+                            client_id: process.env.ONELOGIN_CLIENT_ID!,
+                            client_secret: process.env.ONELOGIN_CLIENT_SECRET!,
+                            redirect_uris: [`${process.env.BACKEND_URL}/auth/onelogin/callback`],
+                            token_endpoint_auth_method: 'private_key_jwt',
+                            token_endpoint_auth_signing_alg: 'PS256',
+                            id_token_signed_response_alg: 'ES256'
+                        },
+                        {
+                            keys: [readPrivateKey(privateKey).export({ format: 'jwk' })]
+                        }
+                    ),
+                    params: {
+                        response_type: 'code',
+                        ui_locales: 'en',
+                        scope: 'openid email'
+                        // TODO: we need to update our OneLogin config to be able to request the below claims
+                        // vtr: ['P2.Cl.Cm'],
+                        // claims: {
+                        //     userinfo: {
+                        //         'https://vocab.account.gov.uk/v1/coreIdentityJWT': null
+                        //     }
+                        // }
+                    }
```

Comment:
Getting a working config for this is what took days of head scratching 😬 

As mentioned in the TODO, we need to bump some settings before we can fetch given and family name.

Reply:
Maybe raise an issue against the TODO here on GitHub so we have some sight of it away from the code and we can come back look again when time or need allows.

Reply:
done